### ISSUE TITLE
Named timers

### ISSUE THREAD

**Issue by jimtng** [boc-tothefuture/openhab-jruby#315]
I'm glad to see #314 and started thinking that it wouldn't be too hard to add a name attribute to timers created using `after`, which can then be used by scripts to keep track of timers. 

In some of my rules, I need to create multiple timers, one for each item, e.g. motion triggers for each room to control its corresponding lights. Each motion sensor would create its own timer.

Currently I'm tracking the timers using a hash with timername => timer_object.

This is a very common usage pattern for me which if added to the core library would be very helpful.

WDYT?

**Comment by ccutrer** [comment]
I do the exact same thing in a few places. Do you have an example of how being able to set a name on the timer would simplify things? I'm having trouble thinking of how it would be any different than a hash with the item as the key.

**Comment by jimtng** [comment]
Here's a contrived / simplistic example:

```
rule 'motion' do
  changed gMotionSensors.members, to: OPEN
  triggered do |sensor|
    light = items[sensor.name.sub(/_Motion$/, '_Light')]
    timer = find_timer(light) # find a timer whose key is the light item
    if timer do
      timer.reschedule
    else
      after(1.minute, light) do |_timer, light| # Assign the light item as the key for this timer
        say "Timer is turning off the #{light.name}"
        light.off
      end
    end
  end
end

rule 'light' do
  changed gLights.members, to: OFF
  triggered do |light|
    find_timer(light)&.cancel # stops the timer from calling "say" when the light is turned off manually
  end
end
```

The key for the timer in this case is not a string, but an actual item object.

**Comment by jimtng** [comment]
> I'm having trouble thinking of how it would be any different than a hash with the item as the key.

It would remove the need to create your own hash (module?) variable to keep track of your timers, at least for straight forward cases.

The syntax can be made more ruby-like perhaps.

**Comment by ccutrer** [comment]
I don't think it's a bad idea at all. I just worry about stuffing the core library with a swiss-army-knife of obscure features that are hard to discover. especially ones that shove methods into the global namespace. This feels like the same category as @boc-tothefuture's trigger attachments idea. almost all of these features can be implemented external to the core library as a module(s) that extends/includes/prepends itself into the core library. See also how I implemented both ensure_states and terse rules. Not core syntax, and I first implemented them as a single thing I could require from my own lib dir, and then they got put into core essentially unchanged from there (and I was a bit hesitant about both of those things being put in this gem, but they seemed to be received well). And given how easy it is to utilize a gem from a rules file, there really shouldn't be any excuse about being able to share little snippets of functionality like this either without having to resort to "download this gist or single file from my repo, and copy it into your own lib dir".

So... I guess if @boc-tothefuture or @pacive think this would be useful generally too, then let's go ahead with it in core, and I'm more than happy to offer advice on how to build it cleanly in ruby.

**Comment by jimtng** [comment]
I share your concern about stuffing things in core, specifically I think `timer` could potentially clash with existing scripts that have already used that name.

As an aside, I am wondering if we could include in the gem a collection of useful extra stuff that need to be `require`d manually. Take this one for example, one could do
```
require 'openhab'
require 'openhab/timer_manager'
```

However, in this case, at least if we go with the syntax above (i.e. `after(duration, lookup_key)`), it would require changing `after`'s syntax a bit. Is there a way to overload a method in ruby so after(duration) (from core) and after(duration, lookup_key) (from `timer_manager`) could co-exist? I suppose it can be implemented using a name other than `after` for the timer_manager version?

**Comment by ccutrer** [comment]
+1 just having it as a separate, optional require. and yes, that's entirely doable. you find whatever class defines `after`, and you basically do this (very much pseudocode):

```
module TimerCreatorWithKeys
  def after(duration, lookup_key, *args)
    timer = super(duration, *args)
    timer.lookup_key = lookup_key
    timer
  end
end

TimerCreator.prepend(TimerCreatorWithKeys)
```

there are several variations though depending on where to hook in. like if it's only a module, and that module has already been included onto Object, all you have to do is Object.include(TimerCreatorWithKeys), which will place your `after` method before the already existing one (assuming Object's `after` method came from an included module or some other ancestor, rather than being defined directly on it). You can always explore these things by looking at object.class.ancestors (in this case, object would be self, aka `main` at the top level), or even better `method(:after)`. `method` returns a `Method` object, that tells you the owner (which module defined it), the source_location, and the `super_method` returns the next `Method` object in the chain.

**Comment by jimtng** [comment]
Am I getting this right: Ruby does method _overriding_ instead of _overloading_ like in c++/java, in order to have a different set of parameters, and it's done by creating a descendent object (hence the call to `super` to invoke the overridden method in the base object). Creating a descendent class/object is done through prepend?

**Comment by jimtng** [comment]
> oh, skimming #16, one thing I should point out is that you don't need to use `$timers` as you do in your examples.

Back then I was weeks old in Ruby.. just trying to convert my jython rules (of which I was a newbie too).



> just using `timers` at the top-level should be good enough in most cases to make that variable visible to all rule executions

I forgot what it was, but something made me use `@@timers` instead of `timers`.

**Comment by jimtng** [comment]
> oh, skimming #16, one thing I should point out is that you don't need to use `$timers` as you do in your examples.

Back then I was weeks old in Ruby.. just trying to convert my jython rules (of which I was a newbie too).



> just using `timers` at the top-level should be good enough in most cases to make that variable visible to all rule executions

I forgot what it was, but something made me use `@@timers` instead of `timers`.

**Comment by boc-tothefuture** [comment]
I added the WIP for reentrant timers in #320 .  See if the example in the docs [here](https://github.com/boc-tothefuture/openhab-jruby/pull/320/files#diff-8f47b636d8e10e25bb6c6769e1b30a61991bde23c092b80a71cb6dd7d4f2cc64R70) solves this use case?

**Comment by jimtng** [comment]
#320 is closer to the solution. There needs to be a way to access a timer from another rule (i.e. Not at the timer creation), ideally to the timer object so we can cancel, check, or reschedule it.

**Comment by jimtng** [comment]
Solved by #320. Thank you, @boc-tothefuture!
